pkg/didcomm/protocol/didexchange/persistence_test.go

```diff
@@ -118,24 +124,161 @@ func TestConnectionRecorder_GetInvitation(t *testing.T) {
+		err := record.saveConnectionRecord(connRec)
 		require.Contains(t, err.Error(), "get error")
 	})
 }
+func TestConnectionRecorder_GetConnectionRecordByProtocolID(t *testing.T) {
+	t.Run(" get connection record by protocol ID in my name space", func(t *testing.T) {
+		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
+		record := NewConnectionRecorder(store)
+		require.NotNil(t, record)
+		connRec := &ConnectionRecord{ThreadID: threadIDValue,
+			ConnectionID: connIDValue, State: stateNameInvited, Namespace: myNSPrefix}
+		err := record.saveNewConnectionRecord(connRec)
+		require.NoError(t, err)
+
+		protocolID, err := createMyNSKey(threadIDValue)
+		require.NoError(t, err)
+
+		storedRecord, err := record.GetConnectionRecordByProtocolID(protocolID)
+		require.NoError(t, err)
+		require.Equal(t, connRec, storedRecord)
+	})
+	t.Run(" get connection record by protocol ID in their name space", func(t *testing.T) {
+		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
+		record := NewConnectionRecorder(store)
+		require.NotNil(t, record)
+		connRec := &ConnectionRecord{ThreadID: threadIDValue,
+			ConnectionID: connIDValue, State: stateNameInvited, Namespace: theirNSPrefix}
+		err := record.saveNewConnectionRecord(connRec)
+		require.NoError(t, err)
+
+		protocolID, err := createTheirNSKey(threadIDValue)
+		require.NoError(t, err)
+
+		storedRecord, err := record.GetConnectionRecordByProtocolID(protocolID)
+		require.NoError(t, err)
+		require.Equal(t, connRec, storedRecord)
+	})
+	t.Run(" get connection record by protocol ID error due to missing input parameter", func(t *testing.T) {
+		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
+		record := NewConnectionRecorder(store)
+		require.NotNil(t, record)
+		connRec, err := record.GetConnectionRecordByProtocolID("")
+		require.Error(t, err)
+		require.Nil(t, connRec)
+	})
+}
+func TestConnectionRecorder_PrepareConnectionRecord(t *testing.T) {
+	t.Run(" prepare connection record  error", func(t *testing.T) {
+		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
+		record := NewConnectionRecorder(store)
+		require.NotNil(t, record)
+		connRec, err := prepareConnectionRecord(nil)
+		require.Error(t, err)
+		require.Nil(t, connRec)
+	})
+}
+func TestConnectionRecorder_CreateNSKeys(t *testing.T) {
+	t.Run(" creating their name space key success", func(t *testing.T) {
+		key, err := createTheirNSKey(threadIDValue)
+		require.NoError(t, err)
+		require.NotNil(t, key)
+	})
+	t.Run(" check error while creating their name space key", func(t *testing.T) {
+		_, err := createTheirNSKey("")
+		require.Error(t, err)
+	})
+	t.Run(" check error while creating my name space key", func(t *testing.T) {
+		_, err := createMyNSKey("")
+		require.Error(t, err)
+	})
+}
+func TestConnectionRecorder_SaveProtocolID(t *testing.T) {
+	t.Run("save protocol id my name space error", func(t *testing.T) {
```

Comment:
you can test the following three scenarios in one test "missing required parameters"

Reply:
done